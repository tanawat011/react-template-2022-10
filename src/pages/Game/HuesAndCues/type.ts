/* istanbul ignore file */
import type { DocumentData, QueryDocumentSnapshot } from 'firebase/firestore'

export type Player = {
  id: string
  name: string
}

export type Room = {
  id: string
  password: string
  hintChoice: string[]
  hintSelected: string
  isStarted: boolean
  isSubmitResult: boolean
  totalRound: number
}

export type RawRoomPlayer = {
  refPlayer: QueryDocumentSnapshot<DocumentData>
  color: string
  score: number
  allSelected: string[]
  isOwner: boolean
  isTurn: boolean
  isHinter: boolean
  totalTurn: number
  seq: number
}

export type RoomPlayer = Omit<RawRoomPlayer, 'refPlayer'> & {
  player: Player
}
