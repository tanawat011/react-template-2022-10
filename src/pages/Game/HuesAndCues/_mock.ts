import type { RoomPlayer } from './_type'

const PLAYER_BASE: RoomPlayer = {
  cells: [],
  uid: '1',
  name: 'Player 1',
  score: 0,
  color: '',
  number: 1,
}

const result = 'F4'
const choice = ['F4', 'A9', 'P27', 'E15']

export const MOCK: {
  ONE_PLAYER: RoomPlayer[]
  TWO_PLAYER: RoomPlayer[]
  TWO_PLAYER_LAST_IS_HINTER: RoomPlayer[]
  THREE_PLAYER: RoomPlayer[]
  THREE_PLAYER_SECOND_IS_HINTER: RoomPlayer[]
  THREE_PLAYER_LAST_IS_HINTER: RoomPlayer[]
  THREE_PLAYER_LAST_IS_TURN: RoomPlayer[]
  HINTER_NO_CHOICE: RoomPlayer[]
  HINTER_NO_SELECT_CHOICE: RoomPlayer[]
  HINTER_SELECTED_CHOICE: RoomPlayer[]
  PLAYER_NO_COLOR: RoomPlayer[]
  PLAYER_HAS_COLOR: RoomPlayer[]
  PLAYER_HAS_COLOR_START_GAME: RoomPlayer[]
  PLAYER_HAS_COLOR_START_GAME_SUBMIT_RESULT: RoomPlayer[]
} = {
  ONE_PLAYER: [
    {
      ...PLAYER_BASE,
      cells: ['D17'],
      number: 1,
      hinter: true,
      isMyTurn: true,
      color: 'bg-[#0758a5]',
    },
  ],
  TWO_PLAYER: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
  ],
  TWO_PLAYER_LAST_IS_HINTER: [
    {
      ...PLAYER_BASE,
      number: 1,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      hinter: true,
    },
  ],
  THREE_PLAYER: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      result,
      choice,
      owner: true,
      score: 1,
    },
    {
      ...PLAYER_BASE,
      uid: '2',
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      uid: '3',
      number: 3,
    },
  ],
  THREE_PLAYER_LAST_IS_TURN: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      result,
      choice,
    },
    {
      ...PLAYER_BASE,
      number: 2,
    },
    {
      ...PLAYER_BASE,
      number: 3,
      isMyTurn: true,
    },
  ],
  THREE_PLAYER_SECOND_IS_HINTER: [
    {
      ...PLAYER_BASE,
      number: 1,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      hinter: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  THREE_PLAYER_LAST_IS_HINTER: [
    {
      ...PLAYER_BASE,
      number: 1,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
      hinter: true,
    },
  ],
  HINTER_NO_CHOICE: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  HINTER_NO_SELECT_CHOICE: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      choice,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  HINTER_SELECTED_CHOICE: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      choice,
      result,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  PLAYER_NO_COLOR: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  PLAYER_HAS_COLOR: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      color: 'bg-[#0758a5]',
      owner: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
      color: 'bg-[#117822]',
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  PLAYER_HAS_COLOR_START_GAME: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      color: 'bg-[#0758a5]',
      owner: true,
      isStartGame: true,
      result: 'F4',
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
      cells: ['F4'],
    },
    {
      ...PLAYER_BASE,
      number: 3,
      cells: ['F5', 'F6'],
    },
  ],
  PLAYER_HAS_COLOR_START_GAME_SUBMIT_RESULT: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      color: 'bg-[#0758a5]',
      owner: true,
      isStartGame: true,
      isSubmitResult: true,
      countRound: 1,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
}
