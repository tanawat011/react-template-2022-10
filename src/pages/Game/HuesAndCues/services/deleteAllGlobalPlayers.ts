import type { RoomPlayer } from '../type'

import { deleteDocument } from 'helpers/firebase'

import { FIRESTORE_PATH } from '../constants'

export const deleteAllGlobalPlayers = async (players: RoomPlayer[]) => {
  const playerPath = FIRESTORE_PATH.DOC_PLAYER

  await Promise.all(
    players.map(async ({ player: { id } }) => {
      const path = playerPath + `/${id}`

      await deleteDocument(path)
    }),
  )
}
