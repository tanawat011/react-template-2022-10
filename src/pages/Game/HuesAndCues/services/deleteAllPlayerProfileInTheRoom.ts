import type { RoomPlayer } from '../_type'

import { getFirestore } from 'firebase/firestore'

import { deleteDocument } from 'helpers/firebase'

import { FIRESTORE_PATH } from '../_constants'

export const deleteAllPlayerProfileInTheRoom = async (players: RoomPlayer[]) => {
  const db = getFirestore()

  await Promise.all(
    players.map(async ({ uid }) => {
      const collectionPath = `${FIRESTORE_PATH.PLAYERS_COLLECTION}/${uid}`

      await deleteDocument(db, collectionPath)
    }),
  )
}
