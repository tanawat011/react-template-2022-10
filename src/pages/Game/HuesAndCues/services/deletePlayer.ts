import { FIRESTORE_PATH } from 'constants/gameHuesAndCues'
import { deleteCollection } from 'helpers/firebase'

export const deletePlayer = async (playerId: string) => {
  const path = FIRESTORE_PATH.DOC_PLAYER + `/${playerId}`

  await deleteCollection(path)
}
