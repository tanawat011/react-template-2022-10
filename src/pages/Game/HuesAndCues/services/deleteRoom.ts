import { getFirestore } from 'firebase/firestore'

import { deleteCollection } from 'helpers/firebase'

import { FIRESTORE_PATH } from '../_constants'

export const deleteRoom = async (roomId: string) => {
  const db = getFirestore()
  const collectionPath = `${FIRESTORE_PATH.ROOMS_DOCUMENT}/${roomId}`

  await deleteCollection(db, collectionPath)
}
