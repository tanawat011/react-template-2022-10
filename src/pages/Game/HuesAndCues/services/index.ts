export * from './deleteAllPlayerProfileInTheRoom'
export * from './deleteRoom'
export * from './getAllPlayersInTheRoom'
export * from './getPlayerInTheRoom'
export * from './getPlayerProfile'
export * from './setPlayerProfile'
export * from './setPlayerInTheRoom'
export * from './subscribeRoom'

export * from './navigate'
