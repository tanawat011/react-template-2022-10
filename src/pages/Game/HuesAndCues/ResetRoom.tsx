import type { RoomPlayer } from './_type'

import { Button } from 'components/Button'

import { getAllPlayersInTheRoom, setPlayerInTheRoom } from './services'

type Prop = {
  roomId: string
  currentPlayer: RoomPlayer
  setCurrentPlayer: (player: RoomPlayer) => void
}

export const ResetRoom: React.FC<Prop> = ({ roomId, currentPlayer, setCurrentPlayer }) => {
  const handleClearGame = async () => {
    try {
      const playersInRoom = await getAllPlayersInTheRoom(roomId)

      playersInRoom.forEach(async (player) => {
        const _player: RoomPlayer = {
          ...player,
          cells: [],
          color: '',
          score: 0,
          result: '',
          choice: [],
          hinter: !!player.owner,
          isSubmitResult: false,
          isStartGame: false,
          countTurn: 0,
          countRound: 0,
        }

        if (currentPlayer.uid === player.uid) {
          setCurrentPlayer(_player)
        }

        await setPlayerInTheRoom(roomId, _player)
      })
    } catch (error) {
      // eslint-disable-next-line no-console
      console.log('error', error)
    }
  }

  return <Button onClick={handleClearGame}>Reset Room</Button>
}
