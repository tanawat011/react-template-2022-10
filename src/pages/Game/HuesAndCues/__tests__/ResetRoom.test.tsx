import type { RenderResult } from '@testing-library/react'

import { fireEvent, screen, act, waitFor } from '@testing-library/react'
import { mocked } from 'jest-mock'

import { renderWithProviders } from 'helpers/test'

import { ResetRoom } from '../ResetRoom'
import { MOCK } from '../_mock'
import { getAllPlayersInTheRoom } from '../services'

jest.mock('components/Button', () => ({
  Button: (props: React.HTMLAttributes<never>) => {
    return (
      <div {...props} data-testid={'button'}>
        {props.children}
      </div>
    )
  },
}))

jest.mock('../services', () => ({
  getAllPlayersInTheRoom: jest.fn(),
  setPlayerInTheRoom: jest.fn(),
}))

describe('<ResetRoom />', () => {
  test('match snapshot', () => {
    const setCurrentPlayer = jest.fn()

    const { asFragment } = renderWithProviders(
      <ResetRoom
        roomId={'room-id-xxx'}
        currentPlayer={MOCK.THREE_PLAYER[0]}
        setCurrentPlayer={setCurrentPlayer}
      />,
    )

    expect(asFragment()).toMatchSnapshot()
  })

  test('clicks reset room', async () => {
    mocked(getAllPlayersInTheRoom).mockResolvedValue(MOCK.THREE_PLAYER)
    const setCurrentPlayer = jest.fn()
    let el: RenderResult

    await act(async () => {
      el = renderWithProviders(
        <ResetRoom
          roomId={'room-id-xxx'}
          currentPlayer={MOCK.THREE_PLAYER[0]}
          setCurrentPlayer={setCurrentPlayer}
        />,
      )
    })

    const button = screen.getByTestId('button')
    fireEvent.click(button)

    await waitFor(() => {
      expect(el.asFragment()).toBeTruthy()
    })
  })

  test('clicks reset room, but error', async () => {
    mocked(getAllPlayersInTheRoom).mockRejectedValue('error')
    const setCurrentPlayer = jest.fn()
    let el: RenderResult

    await act(async () => {
      el = renderWithProviders(
        <ResetRoom
          roomId={'room-id-xxx'}
          currentPlayer={MOCK.THREE_PLAYER[0]}
          setCurrentPlayer={setCurrentPlayer}
        />,
      )
    })

    const button = screen.getByTestId('button')
    fireEvent.click(button)

    await waitFor(() => {
      expect(el.asFragment()).toBeTruthy()
    })
  })
})
