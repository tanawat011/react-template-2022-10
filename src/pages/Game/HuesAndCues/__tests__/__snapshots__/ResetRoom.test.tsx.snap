// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResetRoom /> match snapshot 1`] = `
<DocumentFragment>
  <div
    data-testid="button"
  >
    Reset Room
  </div>
</DocumentFragment>
`;
